js/noms/src/type-cache.js

```diff
@@ -297,84 +320,139 @@ function walkType(t: Type<any>, parentStructTypes: Type<any>[],
+    case Kind.String:
+    case Kind.Type:
+    case Kind.Value:
+      buf.writeUint8(t.kind);
+      break;
+    case Kind.List:
+    case Kind.Map:
+    case Kind.Ref:
+    case Kind.Set:
+      buf.writeUint8(t.kind);
+      buf.writeUint32(desc.elemTypes.length);
+      for (let i = 0; i < desc.elemTypes.length; i++) {
+        encodeForOID(desc.elemTypes[i], buf, allowUnresolvedCycles, root, parentStructTypes);
+      }
+      break;
 
-  if (desc instanceof CycleDesc) {
-    invariant(allowUnresolvedCycles, 'found an unexpected resolved cycle');
-    buf.writeUint8(desc.kind);
-    buf.writeUint32(desc.level);
-  } else if (desc instanceof PrimitiveDesc) {
-    buf.writeUint8(desc.kind);
-  } else if (desc instanceof CompoundDesc) {
-    switch (t.kind) {
-      case Kind.List:
-      case Kind.Map:
-      case Kind.Ref:
-      case Kind.Set: {
-        buf.writeUint8(t.kind);
-        buf.writeUint32(desc.elemTypes.length);
-        for (let i = 0; i < desc.elemTypes.length; i++) {
-          encodeForOID(desc.elemTypes[i], buf, allowUnresolvedCycles, root, parentStructTypes);
-        }
+    case Kind.Union: {
+      buf.writeUint8(t.kind);
+      if (t === root) {
+        // If this is where we started, we don't need to keep going.
         break;
       }
-      case Kind.Union: {
-        buf.writeUint8(t.kind);
-        if (t === root) {
-          // If this is where we started, we don't need to keep going.
-          break;
-        }
-
-        buf.writeUint32(desc.elemTypes.length);
 
-        // This is the only subtle case: encode each subordinate type, generate the hash, remove
-        // duplicates, and xor the results together to form an order indepedant encoding.
-        const mbuf = new BinaryWriter();
-        const oids = new Map();
-        for (let i = 0; i < t.desc.elemTypes.length; i++) {
+      buf.writeUint32(desc.elemTypes.length);
+
+      // This is the only subtle case: encode each subordinate type, generate the hash, remove
+      // duplicates, and xor the results together to form an order indepedant encoding.
+      const mbuf = new BinaryWriter();
+      const oids = new Map();
+      const {elemTypes} = desc;
+      for (let i = 0; i < elemTypes.length; i++) {
+        const elemType = elemTypes[i];
+        let h = getOID(elemType);
+        if (!h) {
           mbuf.reset();
-          encodeForOID(t.desc.elemTypes[i], mbuf, allowUnresolvedCycles, root, parentStructTypes);
-          const h = Hash.fromData(mbuf.data);
-          oids.set(h.toString(), h);
-        }
-
-        const data = alloc(hashByteLength);
-        oids.forEach((oid: Hash) => {
-          const digest = oid.digest;
-          for (let i = 0; i < hashByteLength; i++) {
-            data[i] ^= digest[i];
+          encodeForOID(elemType, mbuf, allowUnresolvedCycles, root, parentStructTypes);
+          h = Hash.fromData(mbuf.data);
+          if (parentStructTypes.indexOf(elemType) === -1) {
+            t.updateOID(h);
```

Comment:
This is wrong! Changing this to elemType and the tests fail. Will have to go back to the drawing board.